core/roboconf-dm/src/main/java/net/roboconf/dm/management/Manager.java

```diff
@@ -955,4 +978,136 @@ void restoreApplications() {
+				long now = System.currentTimeMillis();
+				if ( hasFound || now >= deadline ) {
+					break;
+				}
+
+				// Wait for an Echo message notification!
+				this.echoMessages.wait( deadline - now );
+			}
+		}
+
+		return hasFound;
+	}
+
+
+	/**
+	 * Ping the agent with the specified {@code rootInstanceName} with an
+	 * {@link net.roboconf.messaging.messages.from_dm_to_dm.MsgEcho Echo message}.
+	 *
+	 *
+	 * @param applicationName  the name of the application holding the targeted agent.
+	 * @param rootInstanceName the instance name of the agent to ping.
+	 * @param message          the content of the Echo message to send.
+	 * @param timeout          the timeout in milliseconds (ms) to wait before considering the Echo message is lost.
+	 * @return {@code true} if the sent Echo message has been received before the {@code timeout}, {@code false}
+	 * otherwise.
+	 * @throws java.lang.InterruptedException if interrupted while waiting for the ping.
+	 * @throws java.io.IOException            if something bad happened.
+	 */
+	public boolean pingAgent( String applicationName, String rootInstanceName, String message, long timeout )
+			throws InterruptedException, IOException {
+
+		// Stop 0. Get the application & instance objects.
+		final ManagedApplication ma = this.appNameToManagedApplication.get( applicationName );
+		if (ma == null) {
+			throw new NoSuchElementException( "No application with name: " + applicationName );
+		}
+		final Application app = ma.getApplication();
+		// TODO find the utility method to find the root instance from its name.
+		Instance instance = null;
+		for (Instance i : app.getRootInstances()) {
+			if (i.getName().equals( rootInstanceName )) {
+				instance = i;
+				break;
+			}
+		}
+		if (instance == null) {
+			throw new NoSuchElementException( "No root instance with name " + rootInstanceName + " in application "
+					+ applicationName);
+		}
+
+		// Step 1. Send the PING request message.
+		MsgEcho ping = new MsgEcho( "PING:" + message );
+		this.messagingClient.sendMessageToAgent( app, instance, ping );
+		logger.fine( "Sent PING request message=" + message + "timeout=" + timeout + "ms to application="
+				+ applicationName + ", agent=" + rootInstanceName );
+
+		// Step 2. Wait for the PONG response from the agent.
+		boolean hasFound = false;
+		final long deadline = System.currentTimeMillis() + timeout;
+		final String expectedMessage = "PONG:" + message;
+		while (true) {
+			synchronized ( this.echoMessages ) {
+				// Check all the Echo messages to find our response.
+				for (Iterator<MsgEcho> i = echoMessages.iterator(); i.hasNext(); ) {
+					final MsgEcho m = i.next();
+					if ( m.getContent().equals( expectedMessage ) ) {
+						hasFound = true;
+						i.remove();
+						break;
+					}
+				}
+				long now = System.currentTimeMillis();
+				if ( hasFound || now >= deadline ) {
+					break;
+				}
+
+				// Wait for an Echo message notification!
+				this.echoMessages.wait( deadline - now );
```

Comment:
What happens if the agent never replies?
Does the message remain in the list? When we go through the list, we should clear old messages at the same time. And given the proximity between this code and the code for the DM's debug messages, we could gather it into a single function.